File: components/bass/TagsLibrary/Delphi/Tagging Libraries/WMA Tag Library/Delphi/Tutorial/Unit1.pas
//********************************************************************************************************************************
//*                                                                                                                              *
//*     WMA Tag Library 1.0 © 3delite 2013-2014                                                                                  *
//*     See WMA Tag Library ReadMe.txt for details                                                                               *
//*                                                                                                                              *
//* Two licenses are available for commercial usage of this component:                                                           *
//* Shareware License: €25                                                                                                       *
//* Commercial License: €100                                                                                                     *
//*                                                                                                                              *
//*     http://www.shareit.com/product.html?productid=300579129                                                                  *
//*                                                                                                                              *
//* Using the component in free programs is free.                                                                                *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/WMATagLibrary.html                                         *
//*                                                                                                                              *
//* This component is also available as a part of Tags Library:                                                                  *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/TagsLibrary.html                                           *
//*                                                                                                                              *
//* There is an ID3v2 Library available at:                                                                                      *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/id3v2library.html                                          *
//*                                                                                                                              *
//* and an MP4 Tag Library available at:                                                                                         *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/MP4TagLibrary.html                                         *
//*                                                                                                                              *
//* and an Ogg Vorbis and Opus Tag Library available at:                                                                         *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/OpusTagLibrary.html                                        *
//*                                                                                                                              *
//* and an APEv2 Tag Library available at:                                                                                       *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/APEv2Library.html                                          *
//*                                                                                                                              *
//* and a Flac Tag Library available at:                                                                                         *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/FlacTagLibrary.html                                        *
//*                                                                                                                              *
//* and a WAV Tag Library available at:                                                                                          *
//*                                                                                                                              *
//*     http://www.3delite.hu/Object%20Pascal%20Developer%20Resources/WAVTagLibrary.html                                         *
//*                                                                                                                              *
//* For other Delphi components see the home page:                                                                               *
//*                                                                                                                              *
//*     http://www.3delite.hu/                                                                                                   *
//*                                                                                                                              *
//* If you have any questions or enquiries please mail: 3delite@3delite.hu                                                       *
//*                                                                                                                              *
//* Good coding! :)                                                                                                              *
//* 3delite                                                                                                                      *
//********************************************************************************************************************************

unit Unit1;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, WMATagLibrary, Vcl.StdCtrls, Vcl.ExtCtrls, JPEG, PNGImage, Vcl.ComCtrls;

type
  TForm1 = class(TForm)
    GroupBox1: TGroupBox;
    Button1: TButton;
    Label1: TLabel;
    Edit1: TEdit;
    Button2: TButton;
    OpenDialog1: TOpenDialog;
    GroupBox2: TGroupBox;
    Label2: TLabel;
    Label3: TLabel;
    Label4: TLabel;
    Edit2: TEdit;
    Edit3: TEdit;
    Edit4: TEdit;
    Panel1: TPanel;
    Image1: TImage;
    Button3: TButton;
    Button4: TButton;
    Button5: TButton;
    UpDown1: TUpDown;
    Memo1: TMemo;
    OpenDialog2: TOpenDialog;
    procedure Button2Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure Button4Click(Sender: TObject);
    procedure UpDown1ChangingEx(Sender: TObject; var AllowChange: Boolean; NewValue: SmallInt; Direction: TUpDownDirection);
    procedure Button5Click(Sender: TObject);
    procedure Button3Click(Sender: TObject);
    procedure Button1Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
    CoverArtIndex: Integer;
    procedure DisplayCoverArt(Index: Integer);
  end;

var
  Form1: TForm1;
  WMATag: TWMATag;

implementation

{$R *.dfm}

procedure TForm1.Button1Click(Sender: TObject);
begin
    Close;
end;

procedure TForm1.Button2Click(Sender: TObject);
begin
    if OpenDialog1.Execute then begin
        Edit1.Text := OpenDialog1.FileName;
    end;
end;

procedure TForm1.Button4Click(Sender: TObject);
var
    i: Integer;
begin
    //* Load the tag
    WMATag.LoadFromFile(Edit1.Text);
    //* Display some fields
    Edit2.Text := WMATag.ReadFrameByNameAsText(g_wszWMAuthor);
    Edit3.Text := WMATag.ReadFrameByNameAsText(g_wszWMTitle);
    Edit4.Text := WMATag.ReadFrameByNameAsText(g_wszWMAlbumTitle);
    //* List all attributes
    Memo1.Clear;
    for i := 0 to WMATag.Count - 1 do begin
        if WMATag.Frames[i].Format = WMT_TYPE_STRING then begin
            Memo1.Lines.Add(WMATag.Frames[i].Name + ' = ' + WMATag.Frames[i].GetAsText);
        end else begin
            Memo1.Lines.Add(WMATag.Frames[i].Name + ' = ' + IntToStr(WMATag.Frames[i].GetAsInteger) + ' ('+ IntToStr(WMATag.Frames[i].Stream.Size) + ' bytes)');
        end;
    end;
    //* Display 1st cover art
    DisplayCoverArt(0);
end;

procedure TForm1.Button3Click(Sender: TObject);
begin
    //* Set some fields
    WMATag.SetTextFrameText(g_wszWMAuthor, Edit2.Text);
    WMATag.SetTextFrameText(g_wszWMTitle, Edit3.Text);
    WMATag.SetTextFrameText(g_wszWMAlbumTitle, Edit4.Text);
    //* Note: if value is set it to nothing ('') it will delete the frame
    //* Save the tag
    WMATag.SaveToFile(Edit1.Text);
end;

procedure TForm1.Button5Click(Sender: TObject);
var
    PictureStream: TStream;
    MIMEType: String;
    Description: String;
    Index: Integer;
begin
    if OpenDialog2.Execute then begin
        //* Add a JPEG cover art
        PictureStream := TFileStream.Create(OpenDialog2.FileName, fmOpenRead);
        try
            MIMEType := 'image/jpeg';
            Description := 'No description';
            Index := WMATag.AddFrame(g_wszWMPicture).Index;
            WMATag.SetCoverArtFrame(Index, PictureStream, MIMEType, 3, Description);
        finally
            FreeAndNil(PictureStream);
        end;
    end;
end;

procedure TForm1.FormCreate(Sender: TObject);
begin
    WMATag := TWMATag.Create;
end;

procedure TForm1.FormDestroy(Sender: TObject);
begin
    FreeAndNil(WMATag);
end;

procedure TForm1.UpDown1ChangingEx(Sender: TObject; var AllowChange: Boolean; NewValue: SmallInt; Direction: TUpDownDirection);
begin
    if Direction = updUp then begin
        DisplayCoverArt(CoverArtIndex + 1);
    end;
    if Direction = updDown then begin
        DisplayCoverArt(CoverArtIndex - 1);
    end;
end;

procedure TForm1.DisplayCoverArt(Index: Integer);
var
    ImageJPEG: TJPEGImage;
    ImagePNG: TPNGImage;
    i: Integer;
    CoverArtCounter: Integer;
    MIMEType: String;
    PictureStream: TStream;
    PictureType: Byte;
    Description: String;
begin
    CoverArtCounter := - 1;
    for i := 0 to WMATag.Count - 1 do begin
        if WMATag.Frames[i].Name = g_wszWMPicture then begin
            Inc(CoverArtCounter);
            if CoverArtCounter = Index then begin
                PictureStream := TMemoryStream.Create;
                try
                    if WMATag.GetCoverArtFromFrame(i, PictureStream, MIMEType, PictureType, Description) then begin
                        MIMEType := LowerCase(MIMEType);
                        PictureStream.Seek(0, soBeginning);
                        if (MIMEType = 'image/jpeg')
                        OR (MIMEType = 'image/jpg')
                        then begin
                            ImageJPEG := TJPEGImage.Create;
                            try
                                ImageJPEG.LoadFromStream(PictureStream);
                                Image1.Picture.Assign(ImageJPEG);
                            finally
                                FreeAndNil(ImageJPEG);
                            end;
                        end;
                        if MIMEType = 'image/png' then begin
                            ImagePNG := TPNGImage.Create;
                            try
                                ImagePNG.LoadFromStream(PictureStream);
                                Image1.Picture.Assign(ImagePNG);
                            finally
                                FreeAndNil(ImagePNG);
                            end;
                        end;
                        CoverArtIndex := Index;
                    end;
                finally
                    FreeAndNil(PictureStream);
                end;
            end;
        end;
    end;
end;

end.

File: components/bass/TagsLibrary/Delphi/Tagging Libraries/WMA Tag Library/C++ Builder/Tutorial/Unit1.cpp
//-----------------------------------------------------------------------------
#include <JPEG.hpp>
#include <PNGImage.hpp>
#include <gifimg.hpp>
#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
#include "WMATagLibrary.hpp"

#pragma package(smart_init)
#pragma resource "*.dfm"


TWMATag *WMATag;
int CoverArtIndex;
bool IsSaving;
String LastArtPath;
String LastWMAPath;

//converted to C++ by RMJ 02/2015

TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner) : TForm(Owner)
{
}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button2Click(TObject *Sender)
{
   Edit2->Text = "";
   Edit3->Text = "";
   Edit4->Text = "";
   Memo1->Clear();
   Image1->Picture = NULL;

   OpenDialog1->Filter = L"WM* Files|*.WMV;*.WMA|";
   OpenDialog1->InitialDir = LastWMAPath;
   OpenDialog1->FileName = L"";

   if(OpenDialog1->Execute())
   {
      Edit1->Text = OpenDialog1->FileName;
      LastWMAPath = ExtractFilePath(OpenDialog1->FileName);
   }

   Button4->Click();

}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button4Click(TObject *Sender)
{

   //Load the tag
   WMATag->LoadFromFile(Edit1->Text);
   //Display some fields
   Edit2->Text = WMATag->ReadFrameByNameAsText(g_wszWMAuthor);
   Edit3->Text = WMATag->ReadFrameByNameAsText(g_wszWMTitle);
   Edit4->Text = WMATag->ReadFrameByNameAsText(g_wszWMAlbumTitle);

   //List all attributes
   Memo1->Clear();
   Image1->Picture = NULL;

   for(int i = 0; i < WMATag->Count(); i++)
   {
      if(WMATag->Frames[i]->Format == WMT_TYPE_STRING)
      {
         Memo1->Lines->Add(WMATag->Frames[i]->Name + L" = " + WMATag->Frames[i]->GetAsText());
      }
      else
      {
         Memo1->Lines->Add(WMATag->Frames[i]->Name + L" = " + IntToStr(WMATag->Frames[i]->GetAsInteger()) + " (" + IntToStr(WMATag->Frames[i]->Stream->Size) + " bytes)");
      }
   }

   //Display 1st cover art
   DisplayCoverArt(0);

}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button3Click(TObject *Sender)
{
   String oldCaption = Form1->Caption;
   Form1->Caption = "Saving...please be patient";

   IsSaving = true;
   Button1->Enabled = false;
   Button2->Enabled = false;
   Button3->Enabled = false;
   Button4->Enabled = false;
   Button5->Enabled = false;
   Button7->Enabled = false;
   Button8->Enabled = false;

   //Set some fields
   //Note: if value is set it to nothing L"" it will delete the frame
   WMATag->SetTextFrameText(g_wszWMAuthor, Edit2->Text);
   WMATag->SetTextFrameText(g_wszWMTitle, Edit3->Text);
   WMATag->SetTextFrameText(g_wszWMAlbumTitle, Edit4->Text);

   //Save the tag
   WMATag->SaveToFile(Edit1->Text);

   Button1->Enabled = true;
   Button2->Enabled = true;
   Button3->Enabled = true;
   Button4->Enabled = true;
   Button5->Enabled = true;
   Button7->Enabled = true;
   Button8->Enabled = true;
   Form1->Caption = oldCaption;
   IsSaving = false;

   Button4->Click(); //reload

}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button5Click(TObject *Sender)
{
   TStream *PictureStream;
   String MIMEType, Fext, Description;
   int Index;

   OpenDialog2->Filter = L"Picture Files|*.BMP;*.JPG;*.PNG;*.GIF|";
   OpenDialog2->InitialDir = LastArtPath;
   OpenDialog2->FileName = L"";

   if(OpenDialog2->Execute())
   {
      //Add a cover art
      PictureStream = new TFileStream(OpenDialog2->FileName, fmOpenRead);

      Fext = ExtractFileExt(OpenDialog2->FileName).UpperCase();
      if((Fext == L".JPG") ||(Fext == L".JPEG"))
      {
         MIMEType = L"image/jpg";
      }
      if(Fext == L".PNG")
      {
         MIMEType = L"image/png";
      }
      if(Fext == L".BMP")
      {
         MIMEType = L"image/bmp";
      }
      if(Fext == L".GIF")
      {
         MIMEType = L"image/gif";
      }

      try
      {
         Description = MIMEType; //"No description";
         Index = WMATag->AddFrame(g_wszWMPicture)->Index;
         WMATag->SetCoverArtFrame(Index, PictureStream, MIMEType, 3, Description);
         LastArtPath = ExtractFilePath(OpenDialog2->FileName);

         Button3->Click(); //save
      }
      __finally
      {
         delete PictureStream;
      }

   }

}

//---------------------------------------------------------------------------

void __fastcall TForm1::FormCreate(TObject *Sender)
{

   WMATag = new TWMATag;

}

//---------------------------------------------------------------------------

void __fastcall TForm1::FormDestroy(TObject *Sender)
{
   delete (WMATag);
}

//---------------------------------------------------------------------------

void __fastcall TForm1::DisplayCoverArt(int Index)
{
   TStream    *PictureStream;
   TJPEGImage *ImageJPEG;
   TPngImage  *ImagePNG;
   TGIFImage  *ImageGIF;
   TBitmap    *ImageBMP;
   int i, CoverArtCounter;
   String MIMEType, Description;
   BYTE PictureType;

   CoverArtCounter = -1;

   for(i = 0; i < WMATag->Count(); i++)
   {
      if(WMATag->Frames[i]->Name == g_wszWMPicture)
      {
         CoverArtCounter++;
         if(CoverArtCounter == Index)
         {
            PictureStream = new TMemoryStream;
            try
            {
               if(WMATag->GetCoverArtFromFrame(i, PictureStream, MIMEType, PictureType, Description))
               {
                  MIMEType = LowerCase(MIMEType);
                  PictureStream->Seek((__int64)0, (WORD)soBeginning);
                  if((MIMEType == L"image/jpeg") || (MIMEType == L"image/jpg"))
                  {
                     ImageJPEG = new TJPEGImage;
                     ImageJPEG->LoadFromStream(PictureStream);
                     Image1->Picture->Assign(ImageJPEG);
                     delete (ImageJPEG);
                  }
                  if(MIMEType == L"image/gif")
                  {
                     ImageGIF = new TGIFImage;
                     ImagePNG->LoadFromStream(PictureStream);
                     Image1->Picture->Assign(ImageGIF);
                     delete (ImageGIF);
                  }
                  if(MIMEType == L"image/png")
                  {
                     ImagePNG = new TPngImage;
                     ImagePNG->LoadFromStream(PictureStream);
                     Image1->Picture->Assign(ImagePNG);
                     delete (ImagePNG);
                  }
                  if(MIMEType == L"image/bmp")
                  {
                     ImageBMP = new TBitmap;
                     ImageBMP->LoadFromStream(PictureStream);
                     Image1->Picture->Assign(ImageBMP);
                     delete (ImageBMP);
                  }

                  CoverArtIndex = Index;
               }
            }
            __finally
            {
               delete (PictureStream);
            }

         }

      }

   }

}

//---------------------------------------------------------------------------

void __fastcall TForm1::UpDown1ChangingEx(TObject *Sender, bool &AllowChange, int NewValue, TUpDownDirection Direction)
{

   if(Direction == updUp)
   {
      DisplayCoverArt(CoverArtIndex + 1);
   }
   if(Direction == updDown)
   {
      DisplayCoverArt(CoverArtIndex - 1);
   }

}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button1Click(TObject *Sender)
{

   Close();

}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button7Click(TObject *Sender)
{
    int CoverArtCounter = -1;

   if(Application->MessageBox(L"Are you sure you want to delete the \r\ndisplayed cover from the tag?", L"Confirm",
      MB_YESNO | MB_ICONWARNING | MB_DEFBUTTON1 | MB_APPLMODAL | MB_SETFOREGROUND | MB_TOPMOST) == IDYES)
   {
       //delete the displayed art index
      for(int i = 0; i < WMATag->Count(); i++)
      {
         if(WMATag->Frames[i]->Name == g_wszWMPicture)
         {
            CoverArtCounter++;
            if(CoverArtCounter == CoverArtIndex)
            {
                WMATag->DeleteFrame(i);
                Button3->Click(); //save
                break;
            }
         }
      }
   }
}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button8Click(TObject *Sender)
{
   TStream    *PictureStream;
   TJPEGImage *ImageJPEG;
   TPngImage  *ImagePNG;
   TGIFImage  *ImageGIF;
   TBitmap    *ImageBMP;
   int i, CoverArtCounter;
   String MIMEType, Description, FName;
   BYTE PictureType;

   CoverArtCounter = -1;

   for(i = 0; i < WMATag->Count(); i++)
   {
      if(WMATag->Frames[i]->Name == g_wszWMPicture)
      {
         CoverArtCounter++;
         if(CoverArtCounter == CoverArtIndex)
         {
            PictureStream = new TMemoryStream;

            try
            {
               if(WMATag->GetCoverArtFromFrame(i, PictureStream, MIMEType, PictureType, Description))
               {
                  MIMEType = LowerCase(MIMEType);
                  PictureStream->Seek((__int64)0, (WORD)soBeginning);

                  //JPG
                  if((MIMEType == L"image/jpeg") || (MIMEType == L"image/jpg"))
                  {
                     FName = L"Image.jpg";
                     SaveDialog1->Filter = L"Picture Files|*.JPG|";
                  }
                  //PNG
                  if(MIMEType == L"image/png")
                  {
                     FName = L"Image.png";
                     SaveDialog1->Filter = L"Picture Files|*.PNG|";
                  }
                  //GIF
                  if(MIMEType == L"image/gif")
                  {
                     FName = L"Image.gif";
                     SaveDialog1->Filter = L"Picture Files|*.GIF|";
                  }
                  //BMP
                  if(MIMEType == L"image/bmp")
                  {
                     FName = L"Image.bmp";
                     SaveDialog1->Filter = L"Picture Files|*.BMP|";
                  }

                  SaveDialog1->FileName = FName;

                  if(SaveDialog1->Execute())
                  {
                     if((MIMEType == L"image/jpeg") || (MIMEType == L"image/jpg"))
                     {
                        ImageJPEG = new TJPEGImage;
                        ImageJPEG->LoadFromStream(PictureStream);
                        ImageJPEG->SaveToFile(SaveDialog1->FileName);
                        delete (ImageJPEG);
                     }
                     if(MIMEType == L"image/gif")
                     {
                        ImageGIF = new TGIFImage;
                        ImageGIF->LoadFromStream(PictureStream);
                        ImageGIF->SaveToFile(SaveDialog1->FileName);
                        delete (ImageGIF);
                     }
                     if(MIMEType == L"image/png")
                     {
                        ImagePNG = new TPngImage;
                        ImagePNG->LoadFromStream(PictureStream);
                        ImagePNG->SaveToFile(SaveDialog1->FileName);
                        delete (ImagePNG);
                     }
                     if(MIMEType == L"image/bmp")
                     {
                        ImageBMP = new TBitmap;
                        ImageBMP->LoadFromStream(PictureStream);
                        ImageBMP->SaveToFile(SaveDialog1->FileName);
                        delete (ImageBMP);
                     }

                   }


               }
               else
               {
                  ShowMessage(L"Could not find art data!");
               }

            }
            __finally
            {
               delete (PictureStream);
            }

         }

      }

   }

}

//---------------------------------------------------------------------------

void __fastcall TForm1::FormCloseQuery(TObject *Sender, bool &CanClose)
{

    //try to keep user from corrupting their file
    CanClose = (IsSaving == false);
    if(!CanClose)
      ShowMessage(L"You can not quit while a save is in progress.");

}

//---------------------------------------------------------------------------

void __fastcall TForm1::Button6Click(TObject *Sender)
{
   if(Edit1->Text <= L"")
      return;

   if(Application->MessageBox(L"Are you sure you want to remove the entire tag?\r\n", L"Confirm",
      MB_YESNO|MB_ICONWARNING|MB_DEFBUTTON1|MB_APPLMODAL|MB_SETFOREGROUND|MB_TOPMOST) == IDYES)
   {
      WMATag->DeleteAllFrames();
      WMATag->SaveToFile(Edit1->Text);

      Button4->Click(); //reload
   }

}

//---------------------------------------------------------------------------

